File: PracticaAuxiliar1/2POLIMORFISMO/Mp413/Mp4.py
class MP4:
    def __init__(self, marca, capacidad_gb):
        self.marca = marca
        self.capacidad_gb = capacidad_gb
        self.nro_canciones = 0
        self.nro_videos = 0
        self.canciones = [["" for _ in range(100)] for _ in range(3)]
        self.videos = [["" for _ in range(3)] for _ in range(100)]
    
    def borrar(self, nombre=None, artista=None, peso=None):
        if nombre and not artista and not peso:
            for i in range(self.nro_canciones):
                if self.canciones[0][i] == nombre:
                    self._eliminar_cancion(i)
                    print(f"Canción '{nombre}' eliminada")
                    return
            print(f"Canción '{nombre}' no encontrada")
        
        elif nombre and artista and not peso:
            for i in range(self.nro_canciones):
                if self.canciones[0][i] == nombre and self.canciones[1][i] == artista:
                    self._eliminar_cancion(i)
                    print(f"Canción '{nombre}' de {artista} eliminada")
                    return
            print(f"Canción '{nombre}' de {artista} no encontrada")
        
        elif nombre and artista and peso:
            for i in range(self.nro_canciones):
                if (self.canciones[0][i] == nombre and 
                    self.canciones[1][i] == artista and 
                    self.canciones[2][i] == str(peso)):
                    self._eliminar_cancion(i)
                    print(f"Canción '{nombre}' de {artista} ({peso}KB) eliminada")
                    return
            print("Canción no encontrada con los criterios especificados")
    
    def _eliminar_cancion(self, index):
        for j in range(index, self.nro_canciones - 1):
            self.canciones[0][j] = self.canciones[0][j + 1]
            self.canciones[1][j] = self.canciones[1][j + 1]
            self.canciones[2][j] = self.canciones[2][j + 1]
        self.nro_canciones -= 1
    
    def __add__(self, cancion_info):
        nombre, artista, peso_kb = cancion_info
        if self.nro_canciones >= 100:
            print("No hay espacio para más canciones")
            return self
        
        for i in range(self.nro_canciones):
            if self.canciones[0][i] == nombre and self.canciones[1][i] == artista:
                print("La canción ya existe")
                return self
        
        espacio_necesario = peso_kb / 1024 / 1024
        if self._calcular_espacio_usado() + espacio_necesario > self.capacidad_gb:
            print("No hay suficiente espacio")
            return self
        
        self.canciones[0][self.nro_canciones] = nombre
        self.canciones[1][self.nro_canciones] = artista
        self.canciones[2][self.nro_canciones] = str(peso_kb)
        self.nro_canciones += 1
        print(f"Canción '{nombre}' agregada")
        return self
    
    def __sub__(self, video_info):
        nombre, artista, peso_mb = video_info
        if self.nro_videos >= 100:
            print("No hay espacio para más videos")
            return self
        
        for i in range(self.nro_videos):
            if self.videos[i][0] == nombre and self.videos[i][1] == artista:
                print("El video ya existe")
                return self
        
        espacio_necesario = peso_mb / 1024
        if self._calcular_espacio_usado() + espacio_necesario > self.capacidad_gb:
            print("No hay suficiente espacio")
            return self
        
        self.videos[self.nro_videos][0] = nombre
        self.videos[self.nro_videos][1] = artista
        self.videos[self.nro_videos][2] = str(peso_mb)
        self.nro_videos += 1
        print(f"Video '{nombre}' agregado")
        return self
    
    def _calcular_espacio_usado(self):
        espacio_usado = 0
        
        for i in range(self.nro_canciones):
            if self.canciones[2][i]:
                espacio_usado += float(self.canciones[2][i]) / 1024 / 1024
        
        for i in range(self.nro_videos):
            if self.videos[i][2]:
                espacio_usado += float(self.videos[i][2]) / 1024
        
        return espacio_usado
    
    def mostrar_capacidad(self):
        espacio_usado = self._calcular_espacio_usado()
        espacio_disponible = self.capacidad_gb - espacio_usado
        
        print(f"\n=== CAPACIDAD {self.marca} ===")
        print(f"Capacidad total: {self.capacidad_gb} GB")
        print(f"Espacio usado: {espacio_usado:.4f} GB")
        print(f"Espacio disponible: {espacio_disponible:.4f} GB")
        print(f"Canciones: {self.nro_canciones}")
        print(f"Videos: {self.nro_videos}")
    
    def mostrar_contenido(self):
        print("\n=== CANCIONES ===")
        for i in range(self.nro_canciones):
            print(f"{i+1}. {self.canciones[0][i]} - {self.canciones[1][i]} ({self.canciones[2][i]} KB)")
        
        print("\n=== VIDEOS ===")
        for i in range(self.nro_videos):
            print(f"{i+1}. {self.videos[i][0]} - {self.videos[i][1]} ({self.videos[i][2]} MB)")

if __name__ == "__main__":
    mp4 = MP4("Sony", 1.0)
    
    print("=== EJERCICIO 13 - MP4 ===")
    
    print("\n1. AGREGANDO CONTENIDO:")
    mp4 + ("Back To Black", "Amy Winehouse", 100.0)
    mp4 + ("Lost On You", "LP", 150.0)
    mp4 - ("Heathens", "twenty one pilots", 50.0)
    mp4 - ("Harmonica Andromeda", "KSHMR", 70.0)
    
    print("\n2. MOSTRAR CONTENIDO:")
    mp4.mostrar_contenido()
    
    print("\n3. CAPACIDAD:")
    mp4.mostrar_capacidad()
    
    print("\n4. BORRAR CANCIONES:")
    mp4.borrar("Back To Black")
    mp4.borrar("Lost On You", "LP")
    mp4.borrar("Heathens", "twenty one pilots", "50.0")
    
    print("\n5. CAPACIDAD FINAL:")
    mp4.mostrar_capacidad()

File: PracticaAuxiliar1/2POLIMORFISMO/Mp413/MP4.java
package PracticaAuxiliar1.POLIMORFISMO2.Mp413;

public class MP4 {
    private String marca;
    private double capacidadGb;
    private int nroCanciones;
    private String[][] canciones;
    private int nroVideos;
    private String[][] videos;
    
    public MP4(String marca, double capacidadGb) {
        this.marca = marca;
        this.capacidadGb = capacidadGb;
        this.nroCanciones = 0;
        this.nroVideos = 0;
        this.canciones = new String[3][100];
        this.videos = new String[100][3];
    }
    
    public void borrar(String nombre) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[0][i] != null && canciones[0][i].equals(nombre)) {
                eliminarCancion(i);
                System.out.println("Canción '" + nombre + "' eliminada");
                return;
            }
        }
        System.out.println("Canción '" + nombre + "' no encontrada");
    }
    
    public void borrar(String nombre, String artista) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[0][i] != null && canciones[0][i].equals(nombre) && 
                canciones[1][i] != null && canciones[1][i].equals(artista)) {
                eliminarCancion(i);
                System.out.println("Canción '" + nombre + "' de " + artista + " eliminada");
                return;
            }
        }
        System.out.println("Canción '" + nombre + "' de " + artista + " no encontrada");
    }
    
    public void borrar(String nombre, String artista, String peso) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[0][i] != null && canciones[0][i].equals(nombre) && 
                canciones[1][i] != null && canciones[1][i].equals(artista) &&
                canciones[2][i] != null && canciones[2][i].equals(peso)) {
                eliminarCancion(i);
                System.out.println("Canción '" + nombre + "' de " + artista + " (" + peso + "KB) eliminada");
                return;
            }
        }
        System.out.println("Canción no encontrada con los criterios especificados");
    }
    
    private void eliminarCancion(int index) {
        for (int j = index; j < nroCanciones - 1; j++) {
            canciones[0][j] = canciones[0][j + 1];
            canciones[1][j] = canciones[1][j + 1];
            canciones[2][j] = canciones[2][j + 1];
        }
        canciones[0][nroCanciones - 1] = null;
        canciones[1][nroCanciones - 1] = null;
        canciones[2][nroCanciones - 1] = null;
        nroCanciones--;
    }
    
    public MP4 agregarCancion(String nombre, String artista, double pesoKb) {
        if (nroCanciones >= 100) {
            System.out.println("No hay espacio para más canciones");
            return this;
        }
        
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[0][i] != null && canciones[0][i].equals(nombre) && 
                canciones[1][i] != null && canciones[1][i].equals(artista)) {
                System.out.println("La canción ya existe");
                return this;
            }
        }
        
        double espacioNecesario = pesoKb / 1024 / 1024;
        if (calcularEspacioUsado() + espacioNecesario > capacidadGb) {
            System.out.println("No hay suficiente espacio");
            return this;
        }
        
        canciones[0][nroCanciones] = nombre;
        canciones[1][nroCanciones] = artista;
        canciones[2][nroCanciones] = String.valueOf(pesoKb);
        nroCanciones++;
        System.out.println("Canción '" + nombre + "' agregada");
        return this;
    }
    
    public MP4 agregarVideo(String nombre, String artista, double pesoMb) {
        if (nroVideos >= 100) {
            System.out.println("No hay espacio para más videos");
            return this;
        }
        
        for (int i = 0; i < nroVideos; i++) {
            if (videos[i][0] != null && videos[i][0].equals(nombre) && 
                videos[i][1] != null && videos[i][1].equals(artista)) {
                System.out.println("El video ya existe");
                return this;
            }
        }
        
        double espacioNecesario = pesoMb / 1024;
        if (calcularEspacioUsado() + espacioNecesario > capacidadGb) {
            System.out.println("No hay suficiente espacio");
            return this;
        }
        
        videos[nroVideos][0] = nombre;
        videos[nroVideos][1] = artista;
        videos[nroVideos][2] = String.valueOf(pesoMb);
        nroVideos++;
        System.out.println("Video '" + nombre + "' agregado");
        return this;
    }
    
    public double calcularEspacioUsado() {
        double espacioUsado = 0;
        
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[2][i] != null) {
                espacioUsado += Double.parseDouble(canciones[2][i]) / 1024 / 1024;
            }
        }
        
        for (int i = 0; i < nroVideos; i++) {
            if (videos[i][2] != null) {
                espacioUsado += Double.parseDouble(videos[i][2]) / 1024;
            }
        }
        
        return espacioUsado;
    }
    
    public void mostrarCapacidad() {
        double espacioUsado = calcularEspacioUsado();
        double espacioDisponible = capacidadGb - espacioUsado;
        
        System.out.println("=== CAPACIDAD MP4 ===");
        System.out.println("Marca: " + marca);
        System.out.println("Capacidad total: " + capacidadGb + " GB");
        System.out.println("Espacio usado: " + String.format("%.4f", espacioUsado) + " GB");
        System.out.println("Espacio disponible: " + String.format("%.4f", espacioDisponible) + " GB");
        System.out.println("Número de canciones: " + nroCanciones);
        System.out.println("Número de videos: " + nroVideos);
    }
    
    public void mostrarCanciones() {
        System.out.println("=== CANCIONES ===");
        if (nroCanciones == 0) {
            System.out.println("No hay canciones");
            return;
        }
        for (int i = 0; i < nroCanciones; i++) {
            System.out.println((i + 1) + ". " + canciones[0][i] + " - " + 
                             canciones[1][i] + " (" + canciones[2][i] + " KB)");
        }
    }
    
    public void mostrarVideos() {
        System.out.println("=== VIDEOS ===");
        if (nroVideos == 0) {
            System.out.println("No hay videos");
            return;
        }
        for (int i = 0; i < nroVideos; i++) {
            System.out.println((i + 1) + ". " + videos[i][0] + " - " + 
                             videos[i][1] + " (" + videos[i][2] + " MB)");
        }
    }
}